classes/task/sendfailednotifications_task.php

```diff
@@ -141,9 +148,31 @@ private function sendmail($msg, $subject, $emailto) {
         $from->email = self::DEFAULT_SENDER_EMAIL;
         $from->deleted = 0;
         $from->mailformat = FORMAT_HTML;
-    
-        $messageid = email_to_user($user, $from, $subject, html_to_text($msg), $msg);
-        return $messageid;
+
+        // if any rows haven't been sent before then set flag to send
+        $rowstosend = false;
+        foreach ($failedrows as $row) {
+            $rv = $DB->get_records("logstore_xapi_notif_sent_log", array("failedlogid" => $row->id, "email" => $emailto));
+            if (empty($rv)) {
+                $rowstosend = true;
+                break;
+            }
+        }
```

Comment:
I'm wondering if this is the best way to do this? Say we have 10 failed rows and only 1 of them appears in the sent log for that email address. That would mean the user will get 9 duplicate results in a new email with only 1 being a new result. Would it not make more sense to create a new array of records which have not been sent and use these in the email?

Reply:
This means we would have to build a specific email for each user. The failed log captures the failed statements at that point in time regardless of what was sent to the user before. I think it could be confusing if they logged in and then saw the other failures which had been excluded from their most recent email just because they had received them previously.